CMakeLists.txt

```diff
@@ -167,11 +167,29 @@ if(UNIX)
   message(STATUS "ODBC link flags: ${ODBC_LINK_FLAGS}")
 endif()
 
+
+if(UNIX)
+  find_path(ODBC_INCLUDE sql.h ${find_opt})
+  if(NOT ODBC_INCLUDE)
+    message(FATAL_ERROR "sql.h header file not found")
+  else()
+    message("-- Found sql.h header file at: " ${ODBC_INCLUDE})
+  endif()
+  include_directories(${ODBC_INCLUDE})
+  find_library(ODBC_LIBRARY NAMES odbc PATHS "/usr/local/lib")
+  if(NOT ODBC_LIBRARY)
+    message(FATAL_ERROR "ODBC library not found$")
+  else()
+    message("-- Found ODBC library at: " ${ODBC_LIBRARY})
+  endif()
+  set(lib_dep ${lib_dep} ${ODBC_LIBRARY})
+endif()
```

Comment:
I'm sorry but I'm not convinced to this solution.

I'd rather prefer to switch to CMake 3.12 or later which provides `FindODBC.cmake` module which I submitted a while (https://gitlab.kitware.com/cmake/cmake/-/merge_requests/2069) and is well tested on multiple Unix platforms.

If CMake 3.12 seems too new, then we should simply copy the detection from `FindODBC.cmake` into nanodbc CMake configuration.

By the way, our CMake configuration does need some modernisation in general, and we should strive to make it target-based and avoid commands like `include_directories`.

Reply:
Agreed, the CMake configuration for nanodbc definitely needs to be modernized. I'm not against updating to a newer version of CMake as well.

Reply:
The error that happens with your Cmake version

```
ld: library not found for -lodbc
```

is because you are setting an hard coded library name "odbc".

```
set(ODBCLIB odbc)
```

A possible solution is to use an absolute path for the library name.
Using 

```
find_library(ODBC_LIBRARY NAMES odbc PATHS "/usr/local/lib")
```

it returns an absolute path, the recommended use 

https://cmake.org/cmake/help/v3.2/command/link_directories.html

Reply:
Other solution is to use the path return by

```
odbc_config --libs
-L/usr/local/Cellar/unixodbc/2.3.9/lib -lodbc
```

Reply:
one way to extract the absolute path 

```
if(APPLE)
    # string(SUBSTRING <string> <begin> <length> <output_variable>)
    string(SUBSTRING ${ODBC_LINK_FLAGS} 2 -1 ODBCLIB_tmp) #trim first 3 characters "-L/"
    message(STATUS "ODBCLIB_tmp = ${ODBCLIB_tmp}")
    string(REPLACE " -lodbc" "" ODBCLIB_tmp ${ODBCLIB_tmp})
    message(STATUS "ODBCLIB_tmp = ${ODBCLIB_tmp}")
    set(ODBCLIB "${ODBCLIB_tmp}/libodbc.a")
    message(STATUS "ODBCLIB = ${ODBCLIB}")
  endif()
```

using

https://cmake.org/cmake/help/latest/command/string.html

Reply:
I removed my initial changes and added the above solution

https://github.com/nanodbc/nanodbc/pull/252/commits/032d152fca4123b35ad9966f64b724b735b4a27a

Reply:
After that, one last thing to fix, back to string conversion issues;

```
/Volumes/Home/pvn/dev/nanodbc.pedro-vicente/test/mssql_test.cpp:155:22: error: no matching function for call to 'stoll'
    auto const nid = std::stoll(sid);
                     ^~~~~~~~~~
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/string:4195:37: note: candidate function not viable: no
      known conversion from 'const basic_string<char16_t, char_traits<char16_t>, allocator<char16_t>>' to 'const
      basic_string<char, char_traits<char>, allocator<char>>' for 1st argument
_LIBCPP_FUNC_VIS long long          stoll (const string& __str, size_t* __idx = 0, int __base = 10);
                                    ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/string:4215:37: note: candidate function not viable: no
      known conversion from 'const basic_string<char16_t, char_traits<char16_t>, allocator<char16_t>>' to 'const
      basic_string<wchar_t, char_traits<wchar_t>, allocator<wchar_t>>' for 1st argument
_LIBCPP_FUNC_VIS long long          stoll (const wstring& __str, size_t* __idx = 0, int __base = 10);
```

Reply:
You probably will have a better way to solve that conversion, but this quick conversion works

```
std::wstring tmp;
nanodbc_to_str(tmp, sid);
auto const nid = std::stoll(tmp);
``

where

```
void nanodbc_to_str(std::wstring &dest, const nanodbc::string &src)
{
  for (size_t idx = 0; idx < src.size(); idx++)
  {
    dest += src.at(idx);
  }
}
```

Reply:
ruuning the tests, I get the expected result of "1"  for 

```
REQUIRE(nid == 1);
```

pushed it


https://github.com/nanodbc/nanodbc/pull/252/commits/367ac1a4c335c668de048cb6299905e93ece3a56

Reply:
@pedro-vicente
> The error that happens with your Cmake version
> ```
> ld: library not found for -lodbc
> ```
> is because you are setting an hard coded library relative name "odbc".
> ```
> set(ODBCLIB odbc)
> ```

Please, could you try CMake 3.12 or later with its `FindODBC.cmake`?

> dest += src.at(idx);

This is just 

```cpp
char s;
wchar_t d;
d = s;
```

and it's not the character conversion we want (`std::widen` is not either).

Reply:
> could you try CMake 3.12 or later with its FindODBC.cmake?

FindODBC.cmake found the correct library, but not the right include path;
So, I used it to set the library, and kept some of the old detection for the path
Regarding all string casts (that code was just a  cast for the test), it's better that you decide what to use; pushed this version

https://github.com/pedro-vicente/nanodbc/commit/e3fd3d430a4fe4a6991cde665d2e1dacf83809b5

Reply:
Thanks for testing.

> FindODBC.cmake found the correct library, but not the right include path;

Can you tell what is the right include path?
Was the unixODBC installed from the official packages?

It's important to remember that CMake Find-modules, in general, **are just guessers** and as such they do not promise to accurately detect every library in every environment, and *that* is why  Find-modules provide variables a user can define to workaround such limitations, as I explain here https://gitlab.kitware.com/cmake/cmake/-/merge_requests/2069#note_410663

> The Find*.cmake modules are "guessers" only.
> So, the module has some detection logic for find_module, but users may find it useful to specify the non-standard or not inspected by the module locations via `ODBC_INCLUDE_DIR` and `ODBC_LIBRARY`

Reply:
my bad, I had a missing call to

```
include_directories(${ODBC_INCLUDE_DIRS})
```

I installed with HomeBrew and doing

```
odbc_config --include-prefix        
/usr/local/Cellar/unixodbc/2.3.9/include
```

gives me a different path than the FindODBC module does, but this path also contains the same file

```
-- ODBC include: /usr/local/include
-- ODBC library: /usr/local/lib/libodbc.dylib
-- ODBC config: /usr/local/bin/odbc_config
```

so, now only the call to FindODBC.cmake is present